src/http/http-client.js

```diff
@@ -0,0 +1,218 @@
+
+    /**
+     * Cancels one or more in-progress request(s) by a cancel ID.
+     *
+     * @param {string} cancelId All requests with a matching cancelId
+     *  will be cancelled.
+     */
+    cancel(cancelId) {
+        if (this.cancelTokens[cancelId]) {
+            this.cancelTokens[cancelId].forEach(cancelToken => {
+                cancelToken.cancel(ErrorCodes.USER_CANCELLED);
+            });
+            delete this.cancelTokens[cancelId];
+        }
+        this.cancelledTokens[cancelId] = true;
+    }
+
+    /**
+     * Cancels all in-progress requests, regardless of cancel ID.
+     */
+    cancelAll() {
+        Object.keys(this.cancelTokens).forEach(cancelId => {
+            this.cancel(cancelId);
+        });
+        this.allCancelled = true;
+    }
+
+    /**
+     * Retrieves a value indicating whether a given cancel ID has been
+     * used to cancel an in-progress request. Will return true if the
+     * individual ID was cancelled, or if all requests were cancelled.
+     *
+     * @param {string} cancelId ID to check for cancellation.
+     * @returns {boolean} True if the ID was cancelled, false otherwise.
+     */
+    isCancelled(cancelId) {
+        return this.allCancelled || !!this.cancelledTokens[cancelId];
+    }
+
+    /**
+     * Submits an HTTP request. The client will automatically retry the request if
+     * it encounters a retry-able error (i.e. 5xx errors or network-related errors).
+     * If the request results in a non-success response then the method will throw an
+     * exception containing details of the error.
+     *
+     * The client also provides support for canceling an in-progress request using
+     * its cancel() method.
+     *
+     * Note that all errors that this method throws will be of type UploadError.
+     *
+     * @param {HttpRequest} httpRequest Request to submit. Information will be used to initialize
+     *  an HTTP request. Note that, if present, the request's getCancelId() can be used with
+     *  the cancel() method to cancel this request. In addition, the request will be cancelled
+     *  if the ID matches the name of the file in DirectBinaryUploadController's "cancel" event.
+     *  Note that the ID doesn't need to be unique to a request - meaning that multiple requests
+     *  could be cancelled using a single cancelId.
+     * @param {HttpResult} [httpResult] Optional result to which any retry errors will be
+     *  added.
+     * @returns {HttpResponse} Response to the request, which can be used to retrieve information
+     *  from the response.
+     */
+    async submit(httpRequest, httpResult = false) {
+        const cancelToken = createCancelToken();
+        const tokenId = uuid();
+        registerCancelId.call(this, httpRequest.getCancelId(), tokenId, cancelToken);
+
+        try {
+            return new HttpResponse(this.getOptions(), await timedRequest(httpRequest.toJSON(), getRetryOptions(this.getUploadOptions(), httpResult), cancelToken));
```

Comment:
It's opinionated, but I find it harder to read when having something like `await timedRequest(httpRequest.toJSON()` in function parameters.

A formatting suggestion:
```suggestion
            return new HttpResponse(this.getOptions(), 
                    await timedRequest(httpRequest.toJSON(), 
                    getRetryOptions(this.getUploadOptions(), httpResult), 
                    cancelToken));
```

Reply:
Good callout. That line was too long, so I split it into two lines and applied your formatting suggestion.